ledger/src/blockstore.rs

```diff
@@ -1262,82 +1264,110 @@ impl Blockstore {
         &self,
         shred: &Shred,
         slot_meta: &SlotMeta,
+        just_inserted_data_shreds: &HashMap<(u64, u64), Shred>,
         last_root: &RwLock<u64>,
         leader_schedule: Option<&Arc<LeaderScheduleCache>>,
         is_recovered: bool,
     ) -> bool {
-        let shred_index = u64::from(shred.index());
-        let slot = shred.slot();
-        let last_in_slot = if shred.last_in_slot() {
-            debug!("got last in slot");
-            true
-        } else {
-            false
-        };
+        loop {
+            let shred_index = u64::from(shred.index());
+            let slot = shred.slot();
+            let last_in_slot = if shred.last_in_slot() {
+                debug!("got last in slot");
+                true
+            } else {
+                false
+            };
 
-        // Check that we do not receive shred_index >= than the last_index
-        // for the slot
-        let last_index = slot_meta.last_index;
-        if shred_index >= last_index {
-            let leader_pubkey = leader_schedule
-                .map(|leader_schedule| leader_schedule.slot_leader_at(slot, None))
-                .unwrap_or(None);
-
-            let ending_shred = self.get_data_shred(slot, last_index).unwrap().unwrap();
-            if self
-                .store_duplicate_if_not_existing(slot, ending_shred, shred.payload.clone())
-                .is_err()
-            {
-                warn!("store duplicate error");
+            // Check that we do not receive shred_index >= than the last_index
+            // for the slot
+            let last_index = slot_meta.last_index;
+            if shred_index >= last_index {
+                let leader_pubkey = leader_schedule
+                    .map(|leader_schedule| leader_schedule.slot_leader_at(slot, None))
+                    .unwrap_or(None);
+
+                let ending_shred: Option<Cow<Vec<u8>>> =
+                    if let Some(shred) = just_inserted_data_shreds.get(&(slot, last_index)) {
+                        Some(Cow::Borrowed(&shred.payload))
+                    } else {
+                        self.get_data_shred(slot, last_index)
+                            .unwrap()
+                            .map(Cow::Owned)
+                    };
+
+                if ending_shred.is_none() {
+                    // The slot data might have been cleared between us fetching the SlotMeta and
+                    // retrieving the `ending_shred`. Slot data might have been cleared by
+                    // `Blockstore::clear_unconfirmed_slot()` due to events like duplicate block detection.
+                    // In this case, try to get the updated SlotMeta and try again
```

Comment:
Looks like this is going to need a call to `get_slot_meta_entry` or something similar to actually get updated `SlotMeta`, right? Looping again isn't going to change anything

Reply:
We should probably have an exit hatch for if `SlotMeta` ever gets corrupted such that its `last_index` is still set but the shred doesn't exist

Reply:
Maybe for now we can just have this return an error? This might actually be an invalid shred

Reply:
ha shoot forgot to purge this, it's no longer necessary